### PR TITLE
implement healthchecks - resolves #631

### LINKED ISSUE #631: implement healthchecks for all containers
https://docs.docker.com/engine/reference/builder/#healthcheck

### PR DESCRIPTION
**Pull request by HorayNarea**
- [x] nginx
- [x] dovecot
- [x] postfix
- [x] admin
- [x] roundcube
- [x] rainloop
- [x] rspamd
- [x] clamav
- [x] radicale
- [x] setup
- [x] docs

### CODE REVIEW

**Comment by muhlemmer** [comment]
Tip in advance: on docker Swarm, after a `docker swarm deploy` the docker nodes will start pulling images as required. `depend_on` is not honored. This means containers are started in any random order. And depending on the network connectivity, it can take some time before everything is up. (I've seen case of 3-5 minutes while testing on a cheap-ish host).

During this time some container will remain stuck in `start.py` waiting for resolving to happen. See #560. This means 2 options:
1. Long health check timer for the first poll
2. Use a fake responder in start.py until the main service is up

Maybe there are better solutions, I didn't really check the docks. Just wanted to give the heads-up ;).

**Comment by HorayNarea** [comment]
I wanted this feature more as a quick overview in portainer or with `docker container ls` to see if the services are still responding correctly

the whole `depends_on`-mechanism can later be implemented on top of this

**Comment by kaiyou** [comment]
Healthcheck taking some time before they succeed does not sound like a problem, will it become one?
This looks pretty good to me otherwise.

**Comment by muhlemmer** [comment]
Yes, having failures on it on swarm. I will get bakc later with debugging, as we are busy testing setup.

**Comment by muhlemmer** [comment]
On Docker Swarm, it seem to create an avalanche. This is how it ends up being in Docker Swarm:
```
[root@venus mailu-test]# docker stack ps --no-trunc mailu
ID                          NAME                IMAGE                                                                                            NODE                       DESIRED STATE       CURRENT STATE                 ERROR                                                          PORTS
n3lc6qbqejhod808nwl9bm8fx   mailu_smtp.1        usrpro/postfix:local@sha256:5ca66c3add58493e9c91fbfef7f0a4f87a1b4ca69f22077653c99fc1f6287198     venus.mohlmann.solutions   Running             Starting 26 seconds ago                                                                      
q7c2hefaw958os926rup0kflg   mailu_antispam.1    usrpro/rspamd:local@sha256:5f086f911ed0f65167174694a45d5aba21c0d98f034dd3df6f30198507532e62      venus.mohlmann.solutions   Running             Starting 34 seconds ago                                                                      
4ho949b3klrclqh3fzy86ytiy   mailu_imap.1        usrpro/dovecot:local@sha256:17dc93768a23739bbcca374eef144c54aa5642285635feadde012d7441b86d63     venus.mohlmann.solutions   Running             Starting 38 seconds ago                                                                      
3iz2nglgiv1ghxcct123rwxry   mailu_front.1       usrpro/nginx:local@sha256:44104c43788ea5c748691c6e9fafcdfe2263d2a1d70d8a1f443a78dec8acdb60       venus.mohlmann.solutions   Running             Starting 58 seconds ago                                                                      
2armwl7joa64j4uroop57fma3   mailu_smtp.1        usrpro/postfix:local@sha256:5ca66c3add58493e9c91fbfef7f0a4f87a1b4ca69f22077653c99fc1f6287198     venus.mohlmann.solutions   Shutdown            Failed 31 seconds ago         "task: non-zero exit (137): dockerexec: unhealthy container"   
vfdqyv4q2yi3mu1vcxztucfjy   mailu_antispam.1    usrpro/rspamd:local@sha256:5f086f911ed0f65167174694a45d5aba21c0d98f034dd3df6f30198507532e62      venus.mohlmann.solutions   Shutdown            Failed 39 seconds ago         "task: non-zero exit (137): dockerexec: unhealthy container"   
7eity04rxa1khwz7l6p4zm8iw   mailu_imap.1        usrpro/dovecot:local@sha256:17dc93768a23739bbcca374eef144c54aa5642285635feadde012d7441b86d63     venus.mohlmann.solutions   Shutdown            Failed 43 seconds ago         "task: non-zero exit (137): dockerexec: unhealthy container"   
e1uxmee4zumiic8ydzo8x6vlz   mailu_front.1       usrpro/nginx:local@sha256:44104c43788ea5c748691c6e9fafcdfe2263d2a1d70d8a1f443a78dec8acdb60       venus.mohlmann.solutions   Shutdown            Complete about a minute ago                                                                  
hmhxtlc1pplyr9xtu0lmc78d6   mailu_antivirus.1   usrpro/clamav:local@sha256:565cdbaba9c0c5bb50008b81f5ad6ff74decabaedb3559cccab16bb602449db5      venus.mohlmann.solutions   Running             Running 2 minutes ago                                                                        
nc1wxxc2nikzw71wdqey5hchr   mailu_fetchmail.1   usrpro/fetchmail:local@sha256:0c764395d40053f3f19843118cd1fd13e8b3e0c89babb81ff777087f81987ecd   venus.mohlmann.solutions   Running             Running 4 minutes ago                                                                        
jst1xk6ncov08ne9oxxb6vd9q   mailu_smtp.1        usrpro/postfix:local@sha256:5ca66c3add58493e9c91fbfef7f0a4f87a1b4ca69f22077653c99fc1f6287198     venus.mohlmann.solutions   Shutdown            Failed 2 minutes ago          "task: non-zero exit (137): dockerexec: unhealthy container"   
9xwbom7tx5y8fuihwyq6cxze5   mailu_redis.1       redis:alpine@sha256:2953e537b8eaa5120855285497d4f936d9f02a16480a9d76e8ba014dc3998704             venus.mohlmann.solutions   Running             Running 4 minutes ago                                                                        
rt33rwfy0e1zr2e8okcxscf5b   mailu_admin.1       usrpro/admin:local@sha256:297f0f22d1406f7f09150d81ec808fef0be85698fde219fd7f9eeb6089a900c6       venus.mohlmann.solutions   Running             Running 3 minutes ago                                                                        
w2t51a9v4fyz7ss7ifeyqz9l3   mailu_webmail.1     usrpro/roundcube:local@sha256:b77f6f57f1fc6899ac51ab9bee11c74367edb153d91fb12cdcc03275b6dde05a   venus.mohlmann.solutions   Running             Running 3 minutes ago                                                                        
m97t9z54sojvo4189vcai5x9c   mailu_antispam.1    usrpro/rspamd:local@sha256:5f086f911ed0f65167174694a45d5aba21c0d98f034dd3df6f30198507532e62      venus.mohlmann.solutions   Shutdown            Failed 2 minutes ago          "task: non-zero exit (137): dockerexec: unhealthy container"   
puca0zmf28nexxffrrohbbn69   mailu_webdav.1      usrpro/none:local@sha256:29f930d04517292b97a0455d772d13fdc275a37ac1ac59db0df6c9f12481c6e3        venus.mohlmann.solutions   Running             Running 4 minutes ago                                                                        
yqnpk51xyd9kys4kslmbgvn6z   mailu_imap.1        usrpro/dovecot:local@sha256:17dc93768a23739bbcca374eef144c54aa5642285635feadde012d7441b86d63     venus.mohlmann.solutions   Shutdown            Failed 2 minutes ago          "task: non-zero exit (137): dockerexec: unhealthy container"   
mov24prh2in1l5p14p96y4pte   mailu_front.1       usrpro/nginx:local@sha256:44104c43788ea5c748691c6e9fafcdfe2263d2a1d70d8a1f443a78dec8acdb60       venus.mohlmann.solutions   Shutdown            Complete 2 minutes ago                      
```

For instance on postfix, logs turn out to be empty. An the health-check failure is not very helpfull:

```
[root@venus mailu-test]# docker logs 5cd5bf0a5e8b
[root@venus mailu-test]# docker inspect 5cd5bf0a5e8b
....
            "Health": {
                "Status": "unhealthy",
                "FailingStreak": 3,
                "Log": [
                    {
                        "Start": "2018-10-18T10:09:51.675809348+02:00",
                        "End": "2018-10-18T10:09:52.022893718+02:00",
                        "ExitCode": 1,
                        "Output": ""
                    },
```

So my suspicion is as follows, until a container does not report healthy, the service name remains un-resolvable by other containers. I've included a retry function for that in the past. #560. Probably it remains stuck in the retry phase until docker kills the image.

To confirm, `front` is the primary suspect, as its heath-check is broken on regular docker-compose also. So I enter in the `none` container, install `drill` and try to resolve the `front` hostname. Then I try to resolve the `redis` hostname and the webmail hostname. Redis is without health-check and webmail returns a successful health check.

```
/ # drill front
;; ->>HEADER<<- opcode: QUERY, rcode: NXDOMAIN, id: 62786
;; flags: qr rd ra ; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0 
;; QUESTION SECTION:
;; front.       IN      A

;; ANSWER SECTION:

;; AUTHORITY SECTION:
.       2249    IN      SOA     a.root-servers.net. nstld.verisign-grs.com. 2018101800 1800 900 604800 86400
<omitted lines>
/ # drill redis
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 2360
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; redis.       IN      A

;; ANSWER SECTION:
redis.  600     IN      A       10.0.1.17

;; AUTHORITY SECTION:

<omitted lines>
/ # drill webmail
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 31600
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; webmail.     IN      A

;; ANSWER SECTION:
webmail.        600     IN      A       10.0.1.5

;; AUTHORITY SECTION:

<omitted lines>
```

**Comment by HorayNarea** [comment]
Thanks for the deep dive :grinning:

I will continue working on this, when I have time for it…
Would it be okay for you to continue testing the swarm-stuff after I did more work on it? …because I have almost zero knowledge about swarm :worried:

**Comment by muhlemmer** [comment]
Hi, I'm already halfway a solution since yesterday. Just didn't have time to complete it. Perhaps tonight. You don't have to worry about this one.

On October 21, 2018 4:29:59 PM GMT+03:00, "Thomas Sänger" <notifications@github.com> wrote:
>Thanks for the deep dive :grinning:>
>>
>I will continue working on this, when I have time for it…>
>Would it be okay for you to continue testing the swarm-stuff after I
>did more work on it? …because I have almost zero knowledge about swarm
>:worried:
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Mailu/Mailu/pull/663#issuecomment-431668872

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by muhlemmer** [comment]
See HorayNarea/Mailu#2

**Comment by hacor** [comment]
Sorry to pop in on this subject, but I was wondering, why shouldn't we use just
`rc-status dovecot` or `/etc/init.d/service dovecot status` like explained [here](http://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptact.html) and [here](https://wiki.alpinelinux.org/wiki/Alpine_Linux_Init_System). This way we use the standard implemented checks (if exit code is not 0, somethings wrong)
If I see this correctly, it can be easily implemented for following components:
- Postfix
- Dovecot
- RSPAMD/antivirus

All services providing HTTP service (admin, roundcube,...) We have a simple check to do a curl against the home page and expect a 200 status code.

This way we need as little specific code as possible.
Anyway, I think service testing is very nice and handy!

**Comment by hacor** [comment]
Sorry to pop in on this subject, but I was wondering, why shouldn't we use just
`rc-status dovecot` or `/etc/init.d/service dovecot status` like explained [here](http://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptact.html) and [here](https://wiki.alpinelinux.org/wiki/Alpine_Linux_Init_System). This way we use the standard implemented checks (if exit code is not 0, somethings wrong)
If I see this correctly, it can be easily implemented for following components:
- Postfix
- Dovecot
- RSPAMD/antivirus

All services providing HTTP service (admin, roundcube,...) We have a simple check to do a curl against the home page and expect a 200 status code.

This way we need as little specific code as possible.
Anyway, I think service testing is very nice and handy!

**Comment by hacor** [comment]
Sorry to pop in on this subject, but I was wondering, why shouldn't we use just
`rc-status dovecot` or `/etc/init.d/service dovecot status` like explained [here](http://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptact.html) and [here](https://wiki.alpinelinux.org/wiki/Alpine_Linux_Init_System). This way we use the standard implemented checks (if exit code is not 0, somethings wrong)
If I see this correctly, it can be easily implemented for following components:
- Postfix
- Dovecot
- RSPAMD/antivirus

All services providing HTTP service (admin, roundcube,...) We have a simple check to do a curl against the home page and expect a 200 status code.

This way we need as little specific code as possible.
Anyway, I think service testing is very nice and handy!

File: optional/clamav/Dockerfile [hunk 1]
```diff
@@ -4,8 +4,11 @@ RUN apk add --no-cache clamav rsyslog wget clamav-libunrar
...
 COPY conf /etc/clamav
 COPY start.sh /start.sh
+COPY health.sh /health.sh
 
 EXPOSE 3310/tcp
 VOLUME ["/data"]
 
 CMD ["/start.sh"]
+
+HEALTHCHECK CMD /health.sh
```
Comment (hacor) [inline optional/clamav/Dockerfile]: Couldn't we just use `rc-status clamav` to check it's health? It returns autmatically a nen-zero exit code when somethings wrong

Comment (hacor) [inline optional/clamav/Dockerfile]: Couldn't we just use `rc-status clamav` to check it's health? It returns autmatically a nen-zero exit code when somethings wrong

**Comment by HorayNarea** [comment]
> why shouldn't we use just `rc-status dovecot` or `/etc/init.d/service dovecot status`

because we don't use the init-scripts so they can't tell if the service is running and I wanted the healthchecks to actually test if the service handles connections the way it should do

**Pull request merged**